sql/core/src/test/scala/org/apache/spark/sql/internal/CatalogSuite.scala

```diff
@@ -749,4 +749,22 @@ class CatalogSuite extends SharedSparkSession with AnalysisTest with BeforeAndAf
     assert(spark.catalog.currentCatalog().equals("spark_catalog"))
     assert(spark.catalog.listCatalogs().collect().map(c => c.name).toSet == Set("testcat"))
   }
+
+  test("SPARK-39583: Make RefreshTable be compatible with 3 layer namespace") {
+    val catalogName = "testcat"
+    val dbName = "default"
+    val tableName = "my_table"
+    val tableSchema = new StructType().add("i", "int")
+    val description = "this is a test table"
+    val fullTableName = "testcat.default.my_table"
+
+    createTable(tableName, dbName, catalogName, classOf[FakeV2Provider].getName, tableSchema,
+      Map.empty[String, String], description)
+    spark.catalog.refreshTable(fullTableName)
+    assert(!spark.catalog.isCached(fullTableName))
+
+    spark.catalog.cacheTable(fullTableName)
+    spark.catalog.refreshTable(fullTableName)
+    assert(spark.catalog.isCached(fullTableName))
```

Comment:
@cloud-fan what would be the best way to test `refreshTable`?


I would imagine it is better to 
1. cache the table and fetch the cached data into a local variable
2. change the table
3. refresh the table
4. then compare current cached data from the result in 1.

However I didn't find a good example for how to achieve above steps.

Reply:
We can use `spark.table(...).collect` to fetch the data. If we add/remove files, we can simply check the number of rows.

Reply:
```
    spark.range(10).write.saveAsTable(tableName)
    spark.catalog.cacheTable(tableName)
    assert(spark.table(tableName).collect().length == 10)

    spark.range(5).write.mode(SaveMode.Overwrite)
      .saveAsTable(tableName)
    assert(spark.table(tableName).collect().length == 10)
    spark.catalog.refresh(tableName)
    assert(spark.table(tableName).collect().length == 5)
```

I am doing this but I found `spark.catalog.cacheTable(tableName)` does not take effect as I expect? I expect it will return length == 10 because of the cache. However the second `assert(spark.table(tableName).collect().length == 10)` will just fail to say the test contains 5 rows?